ql/src/java/org/apache/hadoop/hive/ql/QueryCompilationSummaryHook.java

```diff
@@ -0,0 +1,134 @@
+      CompileStep logicalPlan = new CompileStep(PerfLogger.GENERATE_LOGICAL_PLAN);
+      logicalPlanAndOpTree.addSubStep(logicalPlan);
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.PLAN_GENERATION));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.MV_REWRITE_FIELD_TRIMMER));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.REMOVING_SUBQUERY));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.DECORRELATION));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.VALIDATE_QUERY_MATERIALIZATION));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.PREJOIN_ORDERING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.MV_REWRITING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.JOIN_REORDERING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.POSTJOIN_ORDERING));
+      logicalPlan.addSubStep(new CompileStep(PerfLogger.HIVE_SORT_PREDICATES));
+      logicalPlanAndOpTree.addSubStep(new CompileStep(PerfLogger.GENERATE_OPERATOR_TREE));
+      compileSteps.add(logicalPlanAndOpTree);
+      compileSteps.add(new CompileStep(PerfLogger.DEDUCE_RESULTSET_SCHEMA));
+      compileSteps.add(new CompileStep(PerfLogger.PARSE_CONTEXT_GENERATION));
+      compileSteps.add(new CompileStep(PerfLogger.SAVE_AND_VALIDATE_VIEW));
+      compileSteps.add(new CompileStep(PerfLogger.LOGICAL_OPTIMIZATION));
+      compileSteps.add(new CompileStep(PerfLogger.PHYSICAL_OPTIMIZATION));
+      compileSteps.add(new CompileStep(PerfLogger.POST_PROCESSING));
+    }
+
+    public void printCompileSummary() {
+      StringBuilder compileSummary = new StringBuilder();
+      compileSummary.append(QUERY_COMPILATION_SUMMARY);
+      compileSummary.append(System.getProperty(LINE_SEPARATOR));
+      compileSummary.append(SEPARATOR);
+      compileSummary.append(System.getProperty(LINE_SEPARATOR));
+      PerfLogger perfLogger = SessionState.getPerfLogger();
+      appendCompileSteps(compileSummary, perfLogger, "", compileSteps);
+      compileSummary.append(format(TOTAL_COMPILATION_TIME, perfLogger.getDuration(PerfLogger.COMPILE)));
+      compileSummary.append(System.getProperty(LINE_SEPARATOR));
+      compileSummary.append(SEPARATOR);
+      SessionState.getConsole().printInfo(compileSummary.toString(), false);
+    }
+
+    public void appendCompileSteps(StringBuilder compileSummary, PerfLogger perfLogger, String prefix,
+                                   List<CompileStep> currentLevelCompileSteps) {
+      int counter = 0;
+      for (CompileStep compileStep : currentLevelCompileSteps) {
+        counter++;
+        compileSummary.append(format(PerfLogger.COMPILE_STEP + " - " +
+                        prefix + counter + " " + compileStep.name,
+                    perfLogger.getDuration(compileStep.name)));
+        compileSummary.append(System.getProperty(LINE_SEPARATOR));
+        if (compileStep.getSubSteps().isEmpty()) {
+
+          continue;
+        }
```

Comment:
nit: Is this if block necessary? If the subSteps is empty and you pass it recursively to the method then the recursive call will just exit without printing anything.

This is really a minor thing so no need to wait for precommits to run again. If it makes sense you can modify the code just before merging.

Reply:
Done :)